decisions/websre/0005-web-sre-kubernetes-secrets-pattern.md

```diff
@@ -0,0 +1,55 @@
+
+The default assumption for storage patterns in aws secrets store is that you follow the pattern: `<group>/<app>/<environment>/<secretname>` for example `mozmeao/snippets/stage/database_url`  This seems to make writing aws iam access stanzas easy.  Not needing to be rewritten with each new secret added.  But, enough separation to prevent us from accidentally crossing wires.
+
+We want to choose just one secret storage and retrieval method, because having one known good way makes operators experience much simpler.  The alternatives/already existing methods are listed below with a quick note on why we didn't standardize on them.
+
+We are choosing to not use a few other types of secret storage mechanisms, including (but possibly not limited to):
+* git-crypt secret repos + sops, to k8s secrets directly.
+  * Many manual steps to get secrets to prod
+  * Difficult to remove access after people left (git-crypt doesn't support removing keys, you have to re-encrypt)
+* Sealed secrets
+  * The workflow and pattern is similar to externalsecrets, keeping both therefore doesn't make sense (doubles our maintenance burden, with little extra benefit)
+  *  sealed secrets require extra cli tools and setup on operator computers, so is just ever so slightly more difficult to use than externalsecrets
+* SSM Parameter Store (lambdas/codebuilds use this)
+  * is mostly just an earlier version of secrets manager
+  * splitting secrets between two places leads to duplication
+  * the workflows are so similar maintaining both seems sub-optimal
+* no new systems to be built, for example hashicorp vault
+  * that's an entire project in and of itself, so we're choosing to not do that because the 'buy' choice is good enough.
+* adding secrets directly to docker images, or directly in source code (high priority to fix these!)
+  * Avoiding these because security has informed us this is explicitly a bad practice. Makes it way too easy to leak the secret to the world.
+
+
+
+We expect CI/CD pipelines to use their current secret storage, and not try to connect them to the secret store.  The one exception is the cloud native pipeline solutions (e.g AWS CodeBuild), which should use the secrets manager if possible.  We may end up duplicating secrets, but in general we should try to keep application secrets and pipeline secrets separate, with pipeline secrets being manually added in the right context.
+
+## Consequences
+
+- We will be migrating secrets, which is a reasonably risky process for many applications.
+- We will have fewer methods for managing secrets, which ought to make it easier to rotate secrets, keep them up to date, etc.
+- We are gatekeepers for adding new secrets, which may increase unplanned work for our team.
+- If we were to lose the secrets manager completely, rebuilding would be a long and arduous process.
```

Comment:
Similar to my above comment, this feels like something that should be addressed.

There are multiple ways to do this, one pattern here at Mozilla has been to use a sops repo for storage and then push those secrets up to a service. RelOps and RelEng use this method to publish secrets into Vault. 

Alternatively, if the effort for rebuilding is understood and we are willing to accept the risk, then that works as well. My main concern here is that the effort for rebuilding seems to be undefined and potentially long.

Reply:
Yeah, the reason I was listing this in consequences was to try and call the risk out as something to accept.  I think as a team we're willing to accept the risk, it seems really unlikely for AWS to completely lose their secret storage for any length of time, at least under any disaster scenario that isn't really huge.  I think the right tradeoff is to standardize how we feed secrets to applications (while accepting this disaster recovery risk), and then later standardize how we populate secrets (using sops + gcp secrets manager or what have you, or perodic backups or ... ).

Because we're so spread out in terms of current secret solutions, we might be about in that situation today - it's nearly impossible to rebuild the secrets from code. Having the knowledge of how secrets are stored and how to update them is difficult because there are so many answers. Also, quite a few of the applications are out of sync between what's in k8s and what's in the gitcrypt repo. Not as an intentional or design of the system, just a consequence of having so much turnover on the team.

Reply:
My concern would not be AWS or GCP deleting or losing the secrets, but instead a user error or automation error. 

I probably shouldn't have mentioned sops or a specific solution, because that then gets us into comparing the two. I don't think there is anything wrong with this model and I very very very much support standardizing.

>FWIW, all this is a long way of saying that you're not wrong, but it just doesn't seem to be a problem for today, or worth blocking the rest of the work. I'd be curious about if you agree with that statement or not. (you pointed out the problem, but how big of a risk do you view it as, how worried are you about it?)

I think it is a variable risk, which is worrying, but if your team is ok with that then I'm ok with it. I don't think we need to block on it. If there are any specific secrets where the rebuild effort is 1 week+, we might want to consider at least a stop gap backup method for them, but that doesn't need to block this work imo.

Reply:
A potential stopgap is to store such secrets in 1Password vaults, which our team and the dev teams we support are already using for many production secrets.